ntp-daemon/src/ipfilter.rs

```diff
@@ -0,0 +1,336 @@
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+
+use crate::config::subnet::IpSubnet;
+
+/// One part of a BitTree
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq)]
+struct Treenode {
+    // Where in the array the child nodes of this
+    // node are located. A child node is only
+    // generated if the symbol cannot be used to
+    // make a final decision at this level
+    child_offset: u32,
+    inset: u16,
+    outset: u16,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
```

Comment:
I do wonder whether this is all worth it compared to using BTreeMap with (IpAddr, u8).

Reply:
I initially rejected BTree because I didn't see how to get it working (this requires O(n) time with a BTree if you just put in all the ranges, because there can be overlap), but thinking about it now you can do prefiltering on the ranges to get a BTree working as well. It will probably work, but my estimate would be that it would be less efficient, though I am not sure how. Given the need for special construction of the BTree I am not sure this is worth investigating.

Reply:
Presumably if you just need a subnets (that is, the bit filter is only a prefix) than it seems pretty straightforward, I think I've used it before (not sure what you mean by prefiltering).

Reply:
I guess that's true. Adding a check on insertion still seems quite a bit simpler than including your own `BitTree` implementation, though.

Reply:
I agree, and had I thought of that before implementing the current solution, that would have been the route to take. However, I think there is no strong reason at this time to go back and implement another approach given the current state of the code.

Reply:
I could hardly disagree more. @rnijveld any thoughts?

Reply:
I think there's definitely some reasons to go for this rather than something else, but the main one for me would be that we can fully control the efficiency with this implementation, which is something that is definitely important for a deny/allow list given that we want to do that as quickly as possible while using as few resources as possible. Building upon a libstd datastructure cannot give us that level of control.

I think right now we have three options to resolve this

* Complete the Vec-based implementation that @folkertdev started
* Create a BTree based implementation
* Use the implementation from this pull request

I feel that the Vec-based implementation is a long way from being completed, and Folkert already took quite some time to get it to where it is right now. I'm also not too convinced, with all the exceptions that need to be there, it will be much more readable than the implementation in this PR.

Using a BTree would help a little with worst case scenario's compared to the Vec I guess, but I don't see right now how to use the BTreeMap/BTreeSet APIs as efficiently as possible. @djc You wouldn't happen to have any example we could take a look at here? That might help!

The current implementation we have right here might use a custom datastructure, but I feel that the extra control we have here does add some value. Other than that this implementation is already documented, tested, and pretty much is confirmed to be working correctly. The custom datastructure code does take a little while to read, but I don't think it's too excessive to understand, and the API is just as easy as any other implementation would be.

Given all that I think that we should just go ahead and merge this one for now, and maybe at a later point (or if someone really wants to right now) replace this implementation with another one if need be.

@davidv1992 could you do a rebase/merge of main? I don't think there's anything else left to do before a merge, code looks great!

Reply:
@rnijveld Did the rebase